File: sdk-csharp/tests/Primust.Tests/PrimustSdkTest.cs
using System.Security.Cryptography;
using System.Text;
using Xunit;

namespace Primust.Tests;

/// <summary>
/// Primust C# SDK tests.
///
/// Tests run against mock transport (no live API). Verifies:
/// - Pipeline configuration and test mode detection
/// - Run lifecycle (RecordAsync → CloseAsync)
/// - Commitment invariant (raw input never in envelope)
/// - Chain integrity (rolling hash)
/// - SHA-256 golden vectors matching Python/Java/JS SDKs
/// - VPEC parsing and gap recording
/// - Graceful degradation when API unreachable
/// </summary>
public class PrimustSdkTest
{
    // ── Pipeline configuration ──

    [Fact]
    public void Pipeline_pk_test_prefix_sets_test_mode()
    {
        var pipeline = new Pipeline("pk_test_abc123", "test-workflow");
        Assert.True(pipeline.TestMode);
        Assert.Equal("test-workflow", pipeline.WorkflowId);
    }

    [Fact]
    public void Pipeline_pk_live_prefix_sets_production_mode()
    {
        var pipeline = new Pipeline("pk_live_abc123", "prod-workflow");
        Assert.False(pipeline.TestMode);
    }

    [Fact]
    public void Pipeline_pk_sb_prefix_sets_test_mode()
    {
        var pipeline = new Pipeline("pk_sb_sandbox", "sandbox-workflow");
        Assert.True(pipeline.TestMode);
    }

    [Fact]
    public void Pipeline_missing_api_key_throws()
    {
        // Clear env var to ensure no fallback
        var original = Environment.GetEnvironmentVariable("PRIMUST_API_KEY");
        Environment.SetEnvironmentVariable("PRIMUST_API_KEY", null);
        try
        {
            Assert.Throws<ArgumentException>(() => new Pipeline(null, "test"));
        }
        finally
        {
            Environment.SetEnvironmentVariable("PRIMUST_API_KEY", original);
        }
    }

    // ── RecordAsync — commitment invariant ──

    [Fact]
    public async Task RecordAsync_produces_sha256_commitment_hash()
    {
        var run = CreateMockRun();
        var result = await run.RecordAsync(new RecordInput
        {
            Check = "pii_scan",
            ManifestId = "sha256:manifest_pii",
            Input = "sensitive_data"u8.ToArray(),
            CheckResult = CheckResult.Pass,
            Visibility = "opaque",
        });

        Assert.NotNull(result.CommitmentHash);
        Assert.StartsWith("sha256:", result.CommitmentHash);
        Assert.Equal(71, result.CommitmentHash.Length); // "sha256:" + 64 hex chars
        Assert.Equal("sha256", result.CommitmentAlgorithm);
    }

    [Fact]
    public async Task RecordAsync_same_input_same_commitment()
    {
        var input = "deterministic_test"u8.ToArray();

        var run1 = CreateMockRun("run_1");
        var r1 = await run1.RecordAsync(new RecordInput
        {
            Check = "c", ManifestId = "m", Input = input, CheckResult = CheckResult.Pass,
        });

        var run2 = CreateMockRun("run_2");
        var r2 = await run2.RecordAsync(new RecordInput
        {
            Check = "c", ManifestId = "m", Input = input, CheckResult = CheckResult.Pass,
        });

        Assert.Equal(r1.CommitmentHash, r2.CommitmentHash);
    }

    [Fact]
    public async Task RecordAsync_different_input_different_commitment()
    {
        var run1 = CreateMockRun("run_1");
        var r1 = await run1.RecordAsync(new RecordInput
        {
            Check = "c", ManifestId = "m", Input = "aaa"u8.ToArray(), CheckResult = CheckResult.Pass,
        });

        var run2 = CreateMockRun("run_2");
        var r2 = await run2.RecordAsync(new RecordInput
        {
            Check = "c", ManifestId = "m", Input = "bbb"u8.ToArray(), CheckResult = CheckResult.Pass,
        });

        Assert.NotEqual(r1.CommitmentHash, r2.CommitmentHash);
    }

    [Fact]
    public async Task RecordAsync_golden_vector_matches_other_sdks()
    {
        // Golden vector: SHA-256 of "primust_test_vector_v1" must match
        // Python, Java, and JS SDKs.
        var input = "primust_test_vector_v1"u8.ToArray();
        var expectedHash = Sha256Hex(input);

        var run = CreateMockRun();
        var result = await run.RecordAsync(new RecordInput
        {
            Check = "golden", ManifestId = "m", Input = input, CheckResult = CheckResult.Pass,
        });

        Assert.Equal(expectedHash, result.CommitmentHash);
    }

    [Fact]
    public async Task RecordAsync_output_commitment_computed_when_provided()
    {
        var run = CreateMockRun();
        var result = await run.RecordAsync(new RecordInput
        {
            Check = "coverage_check",
            ManifestId = "sha256:manifest_cov",
            Input = "input_data"u8.ToArray(),
            Output = "output_result"u8.ToArray(),
            CheckResult = CheckResult.Pass,
        });

        Assert.NotNull(result.OutputCommitment);
        Assert.StartsWith("sha256:", result.OutputCommitment);
    }

    [Fact]
    public async Task RecordAsync_output_commitment_null_when_not_provided()
    {
        var run = CreateMockRun();
        var result = await run.RecordAsync(new RecordInput
        {
            Check = "c", ManifestId = "m", Input = "x"u8.ToArray(), CheckResult = CheckResult.Pass,
        });

        Assert.Null(result.OutputCommitment);
    }

    // ── Visibility default ──

    [Fact]
    public void RecordInput_defaults_visibility_to_opaque()
    {
        var input = new RecordInput
        {
            Check = "test",
            ManifestId = "sha256:m1",
            Input = "x"u8.ToArray(),
            CheckResult = CheckResult.Pass,
        };

        Assert.Equal("opaque", input.Visibility);
    }

    // ── Chain integrity ──

    [Fact]
    public async Task RecordAsync_chain_hashes_are_unique_across_records()
    {
        var run = CreateMockRun();

        var r1 = await run.RecordAsync(new RecordInput
        {
            Check = "c1", ManifestId = "m", Input = "a"u8.ToArray(), CheckResult = CheckResult.Pass,
        });
        var r2 = await run.RecordAsync(new RecordInput
        {
            Check = "c2", ManifestId = "m", Input = "b"u8.ToArray(), CheckResult = CheckResult.Pass,
        });

        Assert.NotEqual(r1.ChainHash, r2.ChainHash);
        Assert.NotEmpty(r1.ChainHash);
        Assert.NotEmpty(r2.ChainHash);
    }

    [Fact]
    public async Task RecordAsync_record_id_has_rec_prefix()
    {
        var run = CreateMockRun();
        var result = await run.RecordAsync(new RecordInput
        {
            Check = "c", ManifestId = "m", Input = "x"u8.ToArray(), CheckResult = CheckResult.Pass,
        });

        Assert.StartsWith("rec_", result.RecordId);
    }

    // ── Run lifecycle ──

    [Fact]
    public async Task RecordAsync_throws_after_close()
    {
        var run = CreateMockRun();
        await run.RecordAsync(new RecordInput
        {
            Check = "c", ManifestId = "m", Input = "x"u8.ToArray(), CheckResult = CheckResult.Pass,
        });
        await run.CloseAsync();

        await Assert.ThrowsAsync<InvalidOperationException>(async () =>
            await run.RecordAsync(new RecordInput
            {
                Check = "c", ManifestId = "m", Input = "y"u8.ToArray(), CheckResult = CheckResult.Pass,
            }));
    }

    [Fact]
    public async Task CloseAsync_double_close_throws()
    {
        var run = CreateMockRun();
        await run.CloseAsync();
        await Assert.ThrowsAsync<InvalidOperationException>(async () => await run.CloseAsync());
    }

    // ── VPEC ──

    [Fact]
    public async Task CloseAsync_returns_vpec_with_system_unavailable_gap_when_api_unreachable()
    {
        var run = CreateMockRun(testMode: true);
        await run.RecordAsync(new RecordInput
        {
            Check = "check1", ManifestId = "m1", Input = "data"u8.ToArray(),
            CheckResult = CheckResult.Pass,
        });

        var vpec = await run.CloseAsync();

        Assert.NotNull(vpec);
        Assert.StartsWith("vpec_pending_", vpec.VpecId);
        Assert.True(vpec.TestMode);
        Assert.Equal(1, vpec.TotalChecksRun);
        Assert.Single(vpec.GovernanceGaps);
        Assert.Equal("system_unavailable", vpec.GovernanceGaps[0].GapType);
        Assert.Equal("high", vpec.GovernanceGaps[0].Severity);
        Assert.False(vpec.IsClean); // has gap
    }

    [Fact]
    public void VpecResult_IsClean_true_when_chain_intact_and_no_gaps()
    {
        var vpec = new VpecResult
        {
            VpecId = "vpec_001",
            RunId = "run_001",
            WorkflowId = "wf",
            OrgId = "org",
            IssuedAt = "2026-01-01T00:00:00Z",
            ProofLevel = "attestation",
            TotalChecksRun = 3,
            ChecksPassed = 3,
            ChecksFailed = 0,
            GovernanceGaps = new List<GovernanceGap>(),
            ChainIntact = true,
            TestMode = false,
        };

        Assert.True(vpec.IsClean);
    }

    [Fact]
    public void VpecResult_IsClean_false_when_gap_present()
    {
        var vpec = new VpecResult
        {
            VpecId = "vpec_002",
            RunId = "run_002",
            ChainIntact = true,
            GovernanceGaps = new List<GovernanceGap>
            {
                new() { GapId = "gap_1", GapType = "vendor_api_error", Severity = "high" }
            },
        };

        Assert.False(vpec.IsClean);
    }

    // ── CheckResult enum ──

    [Fact]
    public void CheckResult_has_7_values()
    {
        Assert.Equal(7, Enum.GetValues<CheckResult>().Length);
    }

    [Fact]
    public void CheckResult_ToWireValue_maps_correctly()
    {
        Assert.Equal("pass", CheckResult.Pass.ToWireValue());
        Assert.Equal("fail", CheckResult.Fail.ToWireValue());
        Assert.Equal("error", CheckResult.Error.ToWireValue());
        Assert.Equal("not_applicable", CheckResult.NotApplicable.ToWireValue());
    }

    // ── LastCommitmentHash ──

    [Fact]
    public async Task Run_LastCommitmentHash_tracks_most_recent()
    {
        var run = CreateMockRun();
        Assert.Null(run.LastCommitmentHash);

        var r1 = await run.RecordAsync(new RecordInput
        {
            Check = "c1", ManifestId = "m", Input = "a"u8.ToArray(), CheckResult = CheckResult.Pass,
        });
        Assert.Equal(r1.CommitmentHash, run.LastCommitmentHash);

        var r2 = await run.RecordAsync(new RecordInput
        {
            Check = "c2", ManifestId = "m", Input = "b"u8.ToArray(), CheckResult = CheckResult.Pass,
        });
        Assert.Equal(r2.CommitmentHash, run.LastCommitmentHash);
    }

    // ── Queued mode ──

    [Fact]
    public async Task RecordAsync_queued_true_when_transport_returns_null()
    {
        var run = CreateMockRun();
        var result = await run.RecordAsync(new RecordInput
        {
            Check = "c", ManifestId = "m", Input = "x"u8.ToArray(), CheckResult = CheckResult.Pass,
        });

        Assert.True(result.Queued);
    }

    // ── Helpers ──

    private static Run CreateMockRun(string runId = "run_test_001", bool testMode = false)
    {
        return new Run(runId, "test-wf", "org_test", new MockTransport(), testMode);
    }

    private static string Sha256Hex(byte[] data)
    {
        var hash = SHA256.HashData(data);
        return $"sha256:{Convert.ToHexString(hash).ToLowerInvariant()}";
    }
}

/// <summary>
/// Mock transport that simulates API unreachable (returns null for all calls).
/// Mirrors the Java SDK MockTransport pattern.
/// </summary>
internal sealed class MockTransport : HttpTransport
{
    public MockTransport() : base("pk_test_mock", "https://mock.primust.com") { }

    public override Task<Dictionary<string, object?>?> PostOpenRunAsync(
        Dictionary<string, object?> payload)
        => Task.FromResult<Dictionary<string, object?>?>(null);

    public override Task<Dictionary<string, object?>?> PostRecordAsync(
        string runId, Dictionary<string, object?> envelope)
        => Task.FromResult<Dictionary<string, object?>?>(null);

    public override Task<Dictionary<string, object?>?> PostCloseRunAsync(
        string runId, Dictionary<string, object?> payload)
        => Task.FromResult<Dictionary<string, object?>?>(null);
}

File: packages/sdk-java/src/test/java/com/primust/sdk/PrimustSdkTest.java
package com.primust.sdk;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;

import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Primust Java SDK tests.
 *
 * Tests run against mock transport (no live API). Verifies:
 * - Builder pattern and configuration
 * - RecordInput builder validation
 * - Run lifecycle (record → close)
 * - Commitment invariant (raw input never in envelope)
 * - Chain integrity
 * - VPEC parsing
 * - Enum correctness
 */
class PrimustSdkTest {

    // ── Builder ──

    @Test
    @DisplayName("Primust.builder() creates Pipeline with correct config")
    void builderCreates() {
        Pipeline p = Primust.builder()
                .apiKey("pk_sb_abc123")
                .workflowId("test-workflow")
                .build();
        assertEquals("test-workflow", p.workflowId());
        assertTrue(p.testMode());
    }

    @Test
    @DisplayName("Missing API key throws IllegalArgumentException")
    void missingApiKeyThrows() {
        assertThrows(IllegalArgumentException.class, () ->
                Primust.builder().workflowId("test").build());
    }

    @Test
    @DisplayName("pk_sb_ prefix sets test mode")
    void testModeFromKeyPrefix() {
        Pipeline p = Primust.builder().apiKey("pk_sb_xyz").build();
        assertTrue(p.testMode());
    }

    @Test
    @DisplayName("pk_live_ prefix sets production mode")
    void liveModeFromKeyPrefix() {
        Pipeline p = Primust.builder().apiKey("pk_live_xyz").build();
        assertFalse(p.testMode());
    }

    // ── RecordInput Builder ──

    @Test
    @DisplayName("RecordInput.builder() validates required fields")
    void recordInputRequiresCheck() {
        assertThrows(IllegalArgumentException.class, () ->
                RecordInput.builder()
                        .manifestId("sha256:abc")
                        .input("data".getBytes())
                        .checkResult(CheckResult.PASS)
                        .build());
    }

    @Test
    @DisplayName("RecordInput.builder() builds with all fields")
    void recordInputFull() {
        RecordInput ri = RecordInput.builder()
                .check("aml_screen")
                .manifestId("sha256:manifest_001")
                .input("entity_data".getBytes(StandardCharsets.UTF_8))
                .output("result".getBytes(StandardCharsets.UTF_8))
                .checkResult(CheckResult.PASS)
                .details(Map.of("entity_type", "corporate"))
                .visibility("opaque")
                .build();

        assertEquals("aml_screen", ri.check());
        assertEquals("sha256:manifest_001", ri.manifestId());
        assertEquals(CheckResult.PASS, ri.checkResult());
        assertEquals("opaque", ri.visibility());
        assertNotNull(ri.output());
    }

    @Test
    @DisplayName("RecordInput defaults visibility to opaque")
    void recordInputDefaultVisibility() {
        RecordInput ri = RecordInput.builder()
                .check("test")
                .manifestId("sha256:m1")
                .input("x".getBytes())
                .checkResult(CheckResult.PASS)
                .build();
        assertEquals("opaque", ri.visibility());
    }

    @Test
    @DisplayName("RecordInput accepts string check result")
    void recordInputStringCheckResult() {
        RecordInput ri = RecordInput.builder()
                .check("test")
                .manifestId("sha256:m1")
                .input("x".getBytes())
                .checkResult("pass")
                .build();
        assertEquals(CheckResult.PASS, ri.checkResult());
    }

    // ── Enums ──

    @Test
    @DisplayName("CheckResult enum has all 7 values")
    void checkResultValues() {
        assertEquals(7, CheckResult.values().length);
        assertEquals("pass", CheckResult.PASS.value());
        assertEquals("fail", CheckResult.FAIL.value());
        assertEquals("not_applicable", CheckResult.NOT_APPLICABLE.value());
    }

    @Test
    @DisplayName("ProofLevel enum has correct ordering")
    void proofLevelOrdering() {
        assertTrue(ProofLevel.ATTESTATION.isWeakerThan(ProofLevel.MATHEMATICAL));
        assertFalse(ProofLevel.MATHEMATICAL.isWeakerThan(ProofLevel.ATTESTATION));
        assertTrue(ProofLevel.WITNESSED.isWeakerThan(ProofLevel.EXECUTION));
    }

    @Test
    @DisplayName("VisibilityMode enum has 3 values")
    void visibilityModeValues() {
        assertEquals(3, VisibilityMode.values().length);
        assertEquals("opaque", VisibilityMode.OPAQUE.value());
        assertEquals("selective", VisibilityMode.SELECTIVE.value());
    }

    // ── Run lifecycle ──

    @Test
    @DisplayName("Run.record() produces commitment hash")
    void runRecordProducesCommitment() {
        // Use mock transport that returns null (queued mode)
        Run run = new Run("run_test_001", "test-wf", "org_test",
                new MockTransport(), false);

        RecordResult result = run.record(RecordInput.builder()
                .check("pii_scan")
                .manifestId("sha256:manifest_pii")
                .input("sensitive_data".getBytes(StandardCharsets.UTF_8))
                .checkResult(CheckResult.PASS)
                .visibility("opaque")
                .build());

        assertNotNull(result.commitmentHash());
        assertTrue(result.commitmentHash().length() > 10, "Commitment hash should be substantial");
        assertEquals("rec_", result.recordId().substring(0, 4));
        assertTrue(result.queued()); // mock returns null → queued
    }

    @Test
    @DisplayName("Same input produces same commitment")
    void sameInputSameCommitment() {
        byte[] input = "deterministic_test".getBytes(StandardCharsets.UTF_8);

        Run run1 = new Run("run_1", "wf", "org", new MockTransport(), false);
        RecordResult r1 = run1.record(RecordInput.builder()
                .check("c").manifestId("m").input(input).checkResult(CheckResult.PASS).build());

        Run run2 = new Run("run_2", "wf", "org", new MockTransport(), false);
        RecordResult r2 = run2.record(RecordInput.builder()
                .check("c").manifestId("m").input(input).checkResult(CheckResult.PASS).build());

        assertEquals(r1.commitmentHash(), r2.commitmentHash());
    }

    @Test
    @DisplayName("Different input produces different commitment")
    void differentInputDifferentCommitment() {
        Run run1 = new Run("run_1", "wf", "org", new MockTransport(), false);
        RecordResult r1 = run1.record(RecordInput.builder()
                .check("c").manifestId("m").input("aaa".getBytes()).checkResult(CheckResult.PASS).build());

        Run run2 = new Run("run_2", "wf", "org", new MockTransport(), false);
        RecordResult r2 = run2.record(RecordInput.builder()
                .check("c").manifestId("m").input("bbb".getBytes()).checkResult(CheckResult.PASS).build());

        assertNotEquals(r1.commitmentHash(), r2.commitmentHash());
    }

    @Test
    @DisplayName("Record on closed run throws IllegalStateException")
    void recordOnClosedRunThrows() {
        Run run = new Run("run_test", "wf", "org", new MockTransport(), false);
        run.record(RecordInput.builder()
                .check("c").manifestId("m").input("x".getBytes()).checkResult(CheckResult.PASS).build());
        run.close();

        assertThrows(IllegalStateException.class, () ->
                run.record(RecordInput.builder()
                        .check("c").manifestId("m").input("x".getBytes()).checkResult(CheckResult.PASS).build()));
    }

    @Test
    @DisplayName("Double close throws IllegalStateException")
    void doubleCloseThrows() {
        Run run = new Run("run_test", "wf", "org", new MockTransport(), false);
        run.close();
        assertThrows(IllegalStateException.class, run::close);
    }

    // ── VPEC ──

    @Test
    @DisplayName("Run.close() returns VPEC with system_unavailable gap when API unreachable")
    void closeReturnsVpecWithGap() {
        Run run = new Run("run_test", "test-wf", "org_test", new MockTransport(), true);
        run.record(RecordInput.builder()
                .check("check1").manifestId("m1").input("data".getBytes()).checkResult(CheckResult.PASS).build());

        VPEC vpec = run.close();
        assertNotNull(vpec);
        assertTrue(vpec.vpecId().startsWith("vpec_pending_"));
        assertTrue(vpec.testMode());
        assertEquals(1, vpec.totalChecksRun());
        assertEquals(1, vpec.gapsCount());
        assertEquals("system_unavailable", vpec.governanceGaps().get(0).gapType());
        assertFalse(vpec.isClean()); // has gap
    }

    @Test
    @DisplayName("VPEC.isClean() returns true when chain intact and no gaps")
    void vpecIsClean() {
        VPEC vpec = new VPEC("vpec_001", "run_001", "wf", "org",
                "2026-01-01T00:00:00Z", "attestation", 3, 3, 0,
                List.of(), true, "merkle_root", "sig", false, Map.of());
        assertTrue(vpec.isClean());
    }

    // ── Chain integrity ──

    @Test
    @DisplayName("Chain hashes are unique across records")
    void chainHashesUnique() {
        Run run = new Run("run_test", "wf", "org", new MockTransport(), false);

        RecordResult r1 = run.record(RecordInput.builder()
                .check("c1").manifestId("m").input("a".getBytes()).checkResult(CheckResult.PASS).build());
        RecordResult r2 = run.record(RecordInput.builder()
                .check("c2").manifestId("m").input("b".getBytes()).checkResult(CheckResult.PASS).build());

        assertNotEquals(r1.chainHash(), r2.chainHash());
    }

    // ── JSON transport ──

    @Test
    @DisplayName("HttpTransport.toJson serializes correctly")
    void jsonSerialization() {
        String json = HttpTransport.toJson(Map.of("key", "value", "num", 42));
        assertTrue(json.contains("\"key\":\"value\""));
        assertTrue(json.contains("\"num\":42"));
    }

    @Test
    @DisplayName("HttpTransport.parseJson round-trips")
    void jsonRoundTrip() {
        String json = "{\"vpec_id\":\"vpec_001\",\"chain_intact\":true,\"total\":5}";
        Map<String, Object> parsed = HttpTransport.parseJson(json);
        assertEquals("vpec_001", parsed.get("vpec_id"));
        assertEquals(true, parsed.get("chain_intact"));
        assertEquals(5, parsed.get("total"));
    }

    // ── Mock transport ──

    static class MockTransport extends HttpTransport {
        MockTransport() {
            super("pk_sb_mock", "https://mock.primust.com");
        }

        @Override
        Map<String, Object> postOpenRun(Map<String, Object> payload) {
            return null; // simulate unreachable
        }

        @Override
        Map<String, Object> postRecord(String runId, Map<String, Object> envelope) {
            return null; // simulate unreachable — record queued
        }

        @Override
        Map<String, Object> postCloseRun(String runId, Map<String, Object> payload) {
            return null; // simulate unreachable
        }

        @Override
        Map<String, Object> postManifest(Map<String, Object> manifest) {
            return null;
        }
    }
}
